auth/user-store.ts

```diff
@@ -1,27 +1,69 @@
-import { User, UserSchema } from "./user";
+import { User, UserSchema, UserInfo } from "./user";
+import { SocialSignInMock } from "./social-signin-mock";
 
 export class UserStore {
   private nextId = 0;
-  private idStore = new Map<string, UserSchema>();
-  private emailStore = new Map<string, UserSchema>();
+  private store = new Map<string, UserSchema>();
+  private readonly socialMocks: SocialSignInMock[] = [];
 
   add(data: Partial<UserSchema>): User {
     const uid = ++this.nextId + "";
-    const user = new User({ ...data, uid }, this);
+    const user = new User(
+      {
+        ...data,
+        providerData: data.providerId ? [new UserInfo(uid, data.providerId, data)] : [],
+        uid,
+      },
+      this
+    );
 
     const schema = user.toJSON() as UserSchema;
-    this.idStore.set(schema.uid, schema);
-    schema.email && this.emailStore.set(schema.email, schema);
+    this.store.set(schema.uid, schema);
     return user;
   }
 
+  addSocialMock(mock: SocialSignInMock) {
+    this.socialMocks.push(mock);
+  }
+
+  consumeSocialMock(provider: firebase.auth.AuthProvider) {
+    const index = this.socialMocks.findIndex((mock) => mock.type === provider.providerId);
+    if (index === -1) {
+      throw new Error("No mock response set.");
+    }
+
+    // Mock is used, then it must go
+    const mock = this.socialMocks[index];
+    this.socialMocks.splice(index, 1);
+
+    return mock.response;
+  }
+
   findByEmail(email: string): User | undefined {
-    const schema = this.emailStore.get(email);
-    return schema ? new User(schema, this) : undefined;
+    for (const user of this.store.values()) {
+      if (user.email === email) {
+        return new User(user, this);
+      }
+    }
```

Comment:
This seems like it would be a good place for `.find()`

Reply:
I'm quite sure this made more sense when I wrote it 😅

Reply:
The reason why I wrote that that way is because `.values()` returns an iterator. To turn it into an array, one needs to spread it, which is an extra iteration if doing `.find()` on top of it.

Even though that's in theory slower, I went it to make that method shorter 👍